src/box/lua/ctl.c

```diff
@@ -185,6 +187,26 @@ lbox_ctl_wal_sync(struct lua_State *L)
 	return 0;
 }
 
+static int
+lbox_ctl_deactivate_replica(struct lua_State *L)
```

Comment:
I didn't review (I just read it) the RFC and may be missing smth. But why don't we want to remove the replica from the `_cluster` space since we introduce the function, which is called `deactivate_replica`. It seems, judging by its name, the function should do everything, not only deletion of the GC consumer.

We don't want user to touch system spaces at all, and removing replica from the replicaset requires that. Maybe it's time to fix that?

Moreover, I don't like, that it's possible to delete gc consumer for non anonymous replica and allow it to exist in `_cluster` space. It seems to me, that we agreed, that gc consumer must always exist for non-anon replica.

I propose to do that the default behavior: for anon - delete consumer, for non-anon - delete from `_cluster` + delete consumer. Maybe we can introduce the option to `box.ctl.deactivate_replica`, which will say, whether replica should remain in `_cluster`, but frankly speaking, I don't see a reason for such option to exist.

Reply:
I agree. The name `deactivate_replica()` sounds quite generic. Like it must completely disable the replica, no matter what the replica is. Not just delete its `_gc_consumer`. Either the naming is confusing, or the actions must change to complete destruction of the replica, perhaps/

Reply:
> I don't like, that it's possible to delete gc consumer for non anonymous replica and allow it to exist in _cluster space

[Here](https://github.com/tarantool/tarantool/pull/9960#discussion_r1611568006) and [here](https://github.com/tarantool/tarantool/pull/9960#discussion_r1611600335) are comments on motivation of leaving regular replicas without its WAL GC state. It doesn't really matter to me to allow or disallow removing WAL GC state only, but I would add such opportunity since there is a real demand from a real user.

> The name deactivate_replica() sounds quite generic. Like it must completely disable the replica, no matter what the replica is.

Yeah, I agree, we should think about a better name, as you suggested during the design doc review. Do you have any ideas? I could suggest:
* Names associated with WAL: `replica_release_xlogs`/`replica_release_wal`/`replica_clear_wal_gc`.
* Names associated with connection: something like `replica_back_off` - the replica leaves us, but not completely. All such names sound weird somewhy. Also, I came up with `replica_suspend`, but it sounds like "forbid to connect".

> we can introduce the option to `box.ctl.deactivate_replica`

I don't like that the same function can be used to "remove local WAL GC consumer" and to "completely drop the replica and its WAL GC state from all replicas".  These two operations are completely different and have very different scale of consequences.

Reply:
> My suggestion is `drop_gc_consumer`

The function also drops checkpoint reference after checkpoint join.
Also, now we have the only type of WAL GC consumers - `"replica"`. However, in future some new types could be added (at least, consumers are designed in this way), so it would be better to reflect the the function manipulates only replicas.

Reply:
A couple of thoughts about the function:
1. Imagine you need to drop WAL GC state of a replica from the whole replicaset - in order to do this with the new function, you have to call it on all alive replicas and wait for dead ones to call it on them too. Do we have any mechanisms to call a function on all replicas eventually? If we have, will it safe to use them? (For example, the function can be called too late, after a replica appeared and disconnected again).
2. Is there any case when one needs to drop WAL GC state only on one replica?

Reply:
I am joining here that the function's name right now is very confusing. Deactivation can mean many things. Does it delete from `_cluster`? Does it abort the connection on active replicas? When I first saw this function, my expectation was that it disconnects and deletes the replica completely, no matter its state.

I suggest to find a name which is harder to misunderstand. Like `box.ctl.dead_replica_gc()`, `box.ctl.prune_replica()`, `box.ctl.replica_gc()`, or alike. `gc` emphasizes both GC subsystem cleanup and actual deletion of the replica when it has no links anymore (for anon disconnected replicas it is so). `prune` emphasizes that it removes stale replica's state (alongside removing the replica itself if GC was its only state (having an "id" is a state too - this is why non-anon replicas aren't deleted but their state GC info still is dropped)). Maybe there are more names? Again, I find it particularly useful to ask customers what do they think of a name.

Reply:
I like `box.ctl.replica_gc()` and `box.ctl.cleanup_replica()`.

`prune_replica` has strong connotations with removing the replica from _cluster as well. But maybe that's just for me because of an old ticket #3110 (prune dead replicas from a replica set), where we wanted to remove the replicas from _cluster.

I think we shouldn't mention WAL GC consumer in the function name, because we agreed not to mention it in `replication_anon_ttl`. That is, we do not "drop consumer", we "cleanup additional info regarding the replica". Maybe the replica itself as well if it is anonymous.

Reply:
> That is, we do not "drop consumer", we "cleanup additional info regarding the replica"

We cleanup additional info, which is invisible to the user. After "that function" (whatever it's named) is called, replica is still in `box.info.replication`, it's still in `_cluster` space. The only difference from the user side is that now `box.info.gc().consumers` and `box.info.gc().checkpoints` don't mention that replica. So, at least for now, all this function does is removing all gc related stuff. And I suppose, it should mention GC subsystem.

Maybe smth like `replica_drop_gc_state` or just `replica_drop_gc`?

Reply:
> But it should work for anonymous replicas as well, in which case it'll drop the replica, no?

Yep, for anon replica the function drops it from `box.info.replication_anon`

Reply:
> > But it should work for anonymous replicas as well, in which case it'll drop the replica, no?
> 
> Yep, for anon replica the function drops it from `box.info.replication_anon`

That's why I think the function name shouldn't mention wal gc

Reply:
The names proposed by @sergepetrenko are ok for me too. So I suggest Sergey chooses one name and we proceed.